### ISSUE TITLE
Separate virus from other human categories

### ISSUE THREAD

**Issue by claireh93** [ScottishCovidResponse/SCRCIssueTracking#148]
Currently the virus population and human categories (S, E, I, R etc) are stored in the same matrix in the `EpiLandscape`. This probably doesn't make sense anymore, as the update loops for virus and humans act separately anyway. The idea would be to store the virus population in a separate matrix within the `EpiLandscape`. This would require changes to how this landscape is generated and what the update loops act on.

**Comment by claireh93** [comment]
In particular these files would need to be changed, but there may be other places
https://github.com/ScottishCovidResponse/Simulation.jl/blob/dev/src/Epidemiology/EpiLandscape.jl
https://github.com/ScottishCovidResponse/Simulation.jl/blob/dev/src/Epidemiology/EpiGenerate.jl

**Comment by claireh93** [comment]
For the update loops, virusupdate! would now act on a completely separate matrix in `epi.abundances`.
classupdate! would then loop through all classes rather than needing to find where the susceptible class is in lines 113 - 115

**Comment by claireh93** [comment]
In terms of the `EpiLandscape`, it's possible that we could end up with lots of different virus categories in the near future (one per age category for example) - am I right in thinking that @richardreeve?

**Comment by claireh93** [comment]
Blocked until #81 goes through.

**Comment by richardreeve** [comment]
Yes, I think so - I can't see how else we'll manage age specific mixing rates. It also parallelises better the more human and virus states there are, so there's that too...

**Comment by AlexRobson** [comment]
I'm looking at this issue and trying to understand what's involved.
 Is the proposal here to:

1)  add a new matrix to `EpiLandScape` for the viral and human populations. Let's just call it `matrix_h` and `matrix_v` for the sake of discussion: https://github.com/ScottishCovidResponse/Simulation.jl/blob/dev/src/Epidemiology/EpiLandscape.jl#L7

```
mutable struct EpiLandscape
  matrix_h::Matrix{Int64}
  matrix_v::Matrix{Int64} <--- THIS MATRIX ADDED
  grid::Array{Int64, 3}
  seed::Vector{MersenneTwister}
```
2) Update relevant functions such for the virus and human: https://github.com/ScottishCovidResponse/Simulation.jl/blob/dev/src/Epidemiology/EpiGenerate.jl#L49

```
function virusupdate!(epi::EpiSystem, timestep::Unitful.Time)
    dims = _countsubcommunities(epi.epienv.habitat)
...
    Threads.@threads for j in classes
        for i in 1:dims
...
            if epi.epienv.active[x, y]
                traitmatch = traitfun(epi, i, 1)
                # Calculate effective rates
                birthrate = params.virus_growth[j] * timestep * epi.abundances.matrix_v[j, i] <--- THIS MATRIX CHANGED
...
end
```

3) Update classupdate! https://github.com/ScottishCovidResponse/Simulation.jl/blob/dev/src/Epidemiology/EpiGenerate.jl#L102 Now

```
function classupdate!(epi::EpiSystem, timestep::Unitful.Time)
    # Calculate dimenions of habitat and number of classes
    dims = _countsubcommunities(epi.epienv.habitat)
    params = epi.epilist.params
    width = getdimension(epi)[1]
    classes = size(epi.abundances.matrix, 1)
    # Loop through grid squares
    Threads.@threads for i in 1:dims
        rng = epi.abundances.seed[Threads.threadid()]
        susclass = findfirst(epi.epilist.names .== "Susceptible")
        N = sum_pop(epi.abundances.matrix_h, susclass, i) <-- THIS MATRIX CHANGED
        # Loop through classes in chosen square
        for j in susclass:classes
```

I'm a bit confused on this last part though, is the line `susclass = findfirst...` just about finding where the human S, E, I, R categories start for the loop in `for j in susclass:classes`.

**Comment by AlexRobson** [comment]
More broadly is the issue (not this issue, but raised here) in terms of adding new virus classes that `matrix_v` will not be sufficient (we would need (for discussion sake a `matrix_v_18to25`, `matrix_v_26to39` etc as new categories get added. So the design issue in gracefully increasing the number of these matrix compartments. (apologies for any catastrophic misunderstanding here!)

**Comment by claireh93** [comment]
Hi Alex,

> I'm a bit confused on this last part though, is the line `susclass = findfirst...` just about finding where the human S, E, I, R categories start for the loop in `for j in susclass:classes`, as the epi list is by convention always starts with 'Virus': https://github.com/ScottishCovidResponse/Simulation.jl/blob/dev/src/Epidemiology/EpiList.jl...so the change here would be `matrix` -> `matrix_h` and `for j in susclass:classes` -> `for j in 1:classes`?

Yes this is exactly it, so this would allow us to not have to search for where the S class starts and probably save lots of headaches in the future!

>More broadly is the issue (not this issue, but raised here) in terms of adding new virus classes that matrix_v will not be sufficient (we would need (for discussion sake a matrix_v_18to25, matrix_v_26to39 etc as new categories get added. So the design issue in gracefully increasing the number of these matrix compartments. (apologies for any catastrophic misunderstanding here!)

This is an interesting point, I think I was envisioning that the axis of the matrix_v would grow per how many age classes there are. So we would have a M by N matrix, where M is num category and N is num grid cells, for Virus being produced per category. But this would be worth a discussion later today I think!

**Comment by AlexRobson** [comment]
Oh that makes sense. My reading of that then is that It's just different classes of the virus in the way that there are different classes of the human population. So a `matrix_v_<age>` wouldn't be needed, it'd have it's own class label in `EpiList`.

**Comment by AlexRobson** [comment]
Given that the viral and human populations update differently anyway, and (i think) that both need to update the grid location too, would it make sense to do something along these lines:

```
struct Virus <: AbstractAgent end
struct Human <: AbstractAgent end

"""
mutable struct EpiLandscape{T<:AbstractAgent}
  matrix::Matrix{Int64}
  grid::Array{Int64, 3}
  seed::Vector{MersenneTwister}
```

and build `EpiSystem` around:

"""
mutable struct EpiSystem{EE <: AbstractEpiEnv, EL <: EpiList, ER <: AbstractTraitRelationship} <: AbstractEpiSystem{EE, EL, ER}
  abundances::EpiLandscape{Human}
  abundances_v::EpiLandscape{Virus}
"""

Working on this in the initial (rough) proposal I was unsure what the intended logic would be around e.g. `isapprox`, and would lead to duplication of various functions (e.g. pop counting), and so i've come to something like the above. 

I have a WIP here: https://github.com/ScottishCovidResponse/Simulation.jl/tree/ar/issue_148

**Comment by AlexRobson** [comment]
Comment above has been edited quite a lot by me, and it's still a mess :) I will probably edit this too: 

TL;DR: I want to clarify a couple of things: 

1) For the original option [add a new matrix to EpiLandScape], I'm getting a bit stuck on how to update https://github.com/ScottishCovidResponse/Simulation.jl/blob/dev/src/Epidemiology/EpiLandscape.jl#L37 in light of the fact that we are populating two matricies now. 

2) The TL;DR above would be "Would it make sense to introduce some human/virus typing given that the raison d'etre for this is to conceptually split the virus and human abundancy matricies and make an e.g.  EpiLandscape{Human}, EpiLandscape{Virus}, `EpiList{Human}`, `Epilist{Virus}.

**Comment by richardreeve** [comment]
Cool. Thanks for thinking this through and coming up with some options. This may actually tie into the discussion at 2 on Mon about how to represent different interacting factors in the model... weee you thinking of coming to that?

**Comment by claireh93** [comment]
Hi Alex, that's a good point and I hadn't thought about (2) in the first instance, but I think it would be a good direction to head in because it would save a lot of headaches with indexing. Perhaps for now we could put through the smaller changes needed for (1) and think about how (2) might work in the meeting today and following? 

As far as (1) goes, `counttypes` and `countsubcommunities` are functions from a different package Diversity.jl. I think `counttypes` works on epilist.types - so it is specified in e.g. https://github.com/ScottishCovidResponse/Simulation.jl/blob/dev/src/Epidemiology/EpiList.jl#L73 how many types there are. This could be altered for now to not include the virus group? `countsubcommunities` should remain the same though as both human and virus categories are operating over the same number of grid cells.

**Comment by claireh93** [comment]
Sorry I'm not sure whether that totally makes sense, perhaps we can catch up separately at some point to talk all of that through.

**Comment by richardreeve** [comment]
Claire and I can have a chat about this after our separate meeting at 11, but the implications for counttypes may need thinking through depending on how it’s used. Gah. It should be the total number of virus + human types though. My concern is it may be being used to define iterators over type-by-subcommunity matrices, which wouldn’t work now...

**Comment by claireh93** [comment]
Surely it shouldn't be virus + human types if we are having two separate matrices? But yes, we can talk this through later!

**Comment by AlexRobson** [comment]
Unfortunately I can't make that meeting (have another call at that time). Regarding a path going forward to close this i'm happy to focus just on (1) for now, but I may need some guidance on how to handle the highlighted issues (subcommunity and counttypes as they seem like an epi question). Still it would make sense to talk through this rather than github threads.

**Comment by richardreeve** [comment]
Yes. If there's something else you can work on then we can discuss it on Wed? Otherwise let em know and we can find a time to chat...

**Comment by AlexRobson** [comment]
.
